core/trino-main/src/main/java/io/trino/execution/scheduler/BinPackingNodeAllocatorService.java

```diff
@@ -361,37 +466,51 @@ public ReserveResult tryReserve(PendingAcquire acquire)
                         }
                         return false;
                     })
-                    .max(comparing(node -> nodesRemainingMemory.get(node.getNodeIdentifier())));
+                    .collect(toImmutableList());
 
-            if (selectedNode.isEmpty()) {
+            if (candidates.isEmpty()) {
                 return ReserveResult.NONE_MATCHING;
             }
 
-            boolean selectedCandidateMatches = false;
-            if (nodesRemainingMemory.get(selectedNode.get().getNodeIdentifier()) >= acquire.getMemoryLease() || isNodeEmpty(selectedNode.get())) {
+            InternalNode selectedNode = candidates.stream()
+                    .max(comparing(node -> nodesRemainingMemoryRuntimeAdjusted.get(node.getNodeIdentifier())))
+                    .orElseThrow();
+
+            if (nodesRemainingMemoryRuntimeAdjusted.get(selectedNode.getNodeIdentifier()) >= acquire.getMemoryLease() || isNodeEmpty(selectedNode.getNodeIdentifier())) {
```

Comment:
Should we introduce a "headroom" here and avoid scheduling if the actual memory (runtime based) is, say 90% utilized?

Reply:
Per offline discussion added separate commit allowing per-task runtime memory estimation overhead (default set to 1GB).